include/mbedtls/ecdsa.h

```diff
@@ -51,9 +51,40 @@
 #if MBEDTLS_ECP_MAX_BYTES > 124
 #error "MBEDTLS_ECP_MAX_BYTES bigger than expected, please fix MBEDTLS_ECDSA_MAX_LEN"
 #endif
-/** The maximal size of an ECDSA signature in Bytes. */
-#define MBEDTLS_ECDSA_MAX_LEN  ( 3 + 2 * ( 3 + MBEDTLS_ECP_MAX_BYTES ) )
 
+/**
+ * \brief           Maximum ECDSA signature size for a given curve bit size
+ *
+ * \param bits      Curve size in bits
+ * \return          Maximum signature size in bytes
+ *
+ * \note            This macro returns a compile-time constant if its argument
+ *                  is one. It may evaluate its argument multiple times; if
+ *                  this is a problem, call the function
+ *                  mbedtls_ecdsa_max_sig_len instead.
+ */
+#define MBEDTLS_ECDSA_MAX_SIG_LEN( bits )                               \
+    ( /*T,L of SEQUENCE*/ ( ( bits ) >= 61 * 8 ? 3 : 2 ) +              \
```

Comment:
@gilles-peskine-arm @mpg How does this come about? As far as I see, we have 1-byte length encodings for sizes <= 127 bytes, so if the bitsize is 61 bytes, the MPI encoding of `r,s` would be `61 + 2 = 63` each, so the `SEQUENCE` content would be `126`, which could still be tagged and length-encoded with 2 bytes. So, cutting the question short: Why is it `( bits ) >= 61 * 8` here instead of `( bits ) >= 62 * 8`?

Reply:
You left out the leading 0 bit in the ASN.1 bignum encoding.

Experimental proof:
```
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include "mbedtls/config.h"
#include "mbedtls/bignum.h"
#include "mbedtls/ecdsa.h"
#include "mbedtls/error.h"

#define CHECK( ret )                                                    \
    do {                                                                \
        int CHECK__ret = ( ret );                                       \
        char CHECK__msg [100];                                          \
        if( CHECK__ret != 0 )                                           \
        {                                                               \
            mbedtls_strerror( CHECK__ret, CHECK__msg, sizeof( CHECK__msg ) ); \
            fprintf( stderr, "%s:%d: %s\n  in %s\n",                    \
                     __FILE__, __LINE__, CHECK__msg, #ret );            \
            exit( 1 );                                                  \
        }                                                               \
    } while( 0 )

int main( int argc, char *argv[] )
{
    mbedtls_mpi r, s;
    unsigned char sig[MBEDTLS_ECDSA_MAX_LEN];
    size_t sig_len;
    int bits;
    mbedtls_mpi_init( &r );
    mbedtls_mpi_init( &s );
    bits = atoi( argv[1] );
    CHECK( mbedtls_mpi_lset( &r, 1 ) );
    CHECK( mbedtls_mpi_lset( &s, 1 ) );
    CHECK( mbedtls_mpi_shift_l( &r, bits - 1 ) );
    CHECK( mbedtls_mpi_shift_l( &s, bits - 1 ) );
    CHECK( mbedtls_ecdsa_signature_to_asn1( &r, &s,
                                            sig, &sig_len, sizeof( sig ) ) );
    fprintf( stderr, "bits=%d => MAX_SIG_LEN = %u, sig_len = %zu\n",
             bits, MBEDTLS_ECDSA_MAX_SIG_LEN( bits ), sig_len );
    fwrite( sig, 1, sig_len, stdout );
    return( 0 );
}
```
Run:
```
for x in 60 61 62 63 64; do ./a.out $(($x*8)) | od -t x1; done
```

Reply:
Thanks for the explanation!